src/DeviceVerificationFlow.cpp

```diff
@@ -0,0 +1,519 @@
+                this,
+                [this](const mtx::events::collections::DeviceEvents &message) {
+                        auto msg =
+                          std::get<mtx::events::DeviceEvent<msgs::KeyVerificationMac>>(message);
+                        if (msg.content.transaction_id == this->transaction_id) {
+                                std::string info =
+                                  "MATRIX_KEY_VERIFICATION_MAC" + this->toClient.to_string() +
+                                  this->deviceId.toStdString() +
+                                  http::client()->user_id().to_string() +
+                                  http::client()->device_id() + this->transaction_id;
+
+                                std::vector<std::string> key_list;
+                                std::string key_string;
+                                for (auto mac : msg.content.mac) {
+                                        if (mac.second ==
+                                            this->sas->calculate_mac(this->device_keys[mac.first],
+                                                                     info + mac.first)) {
+                                                key_string += mac.first;
+                                        } else {
+                                                this->cancelVerification();
+                                                return;
+                                        }
+                                }
+                                if (msg.content.keys ==
+                                    this->sas->calculate_mac(key_string, info + "KEY_IDS")) {
+                                        this->sendVerificationDone();
+                                } else {
+                                        this->cancelVerification();
+                                }
+                        }
+                });
+        connect(ChatPage::instance(),
+                &ChatPage::recievedDeviceVerificationReady,
+                this,
+                [this](const mtx::events::collections::DeviceEvents &message) {
+                        auto msg =
+                          std::get<mtx::events::DeviceEvent<msgs::KeyVerificationReady>>(message);
+                        if (msg.content.transaction_id == this->transaction_id) {
+                                this->startVerificationRequest();
+                        }
+                });
+        connect(ChatPage::instance(),
+                &ChatPage::recievedDeviceVerificationDone,
+                this,
+                [this](const mtx::events::collections::DeviceEvents &message) {
+                        auto msg =
+                          std::get<mtx::events::DeviceEvent<msgs::KeyVerificationDone>>(message);
+                        if (msg.content.transaction_id == this->transaction_id) {
+                                this->startVerificationRequest();
+                                emit this->deviceVerified();
```

Comment:
If this does not arrive in all cases, can you comment on this MSC, that not all clients currently send this? https://github.com/matrix-org/matrix-doc/pull/2366#discussion_r426163446

Reply:
Currently no client supports it I have tried it with both with riot-web and riot-android, so for now ig I will comment it.